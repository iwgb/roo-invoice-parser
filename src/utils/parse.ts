import chunk from 'lodash/chunk';
import { DateTime, DateTimeOptions, Info } from 'luxon';
import { Adjustment, Shift } from '../types';
import { getDateTime } from './datetime';
import { INVOICE_DATE_FORMAT } from '../constants/invoice';

const INVOICE_TIME_REGEX = /^\d\d:\d\d$/;
const INVOICE_DATE_REGEX = /\d{1,2} \w+ \d{4}/;

const getDataFromShiftLine = (shiftLine: string[], opts: DateTimeOptions) => {
  const dataStartIndex = shiftLine.findIndex((item) => INVOICE_TIME_REGEX.test(item));
  const dateMatches = shiftLine
    .slice(0, dataStartIndex)
    .join(' ')
    .match(INVOICE_DATE_REGEX);
  const date = dateMatches === null
    ? ''
    : dateMatches[0];

  const [startTime, endTime, _hours, ordersAndTotal] = shiftLine.slice(dataStartIndex);
  const [orders, total] = ordersAndTotal.split(':');

  const start = getDateTime(date, startTime, opts);
  let end = getDateTime(date, endTime, opts);

  if (end < start) {
    end = end.plus({ days: 1 });
  }

  return {
    start,
    end,
    orders: Number.parseInt(orders, 10),
    pay: Math.round(Number.parseFloat(total.trim().slice(1)) * 100),
  };
};

export const getDataFromShiftTable = (
  text: string[],
  zone: string,
  locale: string,
  lineBefore: string,
  lineAfter: string,
  additionalLineStartFlags: string[] = [],
): Omit<Shift, 'hours'>[] => {
  const weekdays = [
    ...Info.weekdays('long', { locale })
      .map((weekday) => weekday.toLowerCase()),
    ...additionalLineStartFlags,
  ];

  const rawShifts = text
    .slice(
      text.indexOf(lineBefore) + 1,
      text.indexOf(lineAfter),
    )
    .reduce((shifts, shiftLine) => {
      const lastShift = shifts.slice(-1)[0] || [];
      const isNewRow = weekdays.includes(shiftLine.toLowerCase());

      if (isNewRow) {
        return [
          ...shifts,
          [shiftLine],
        ];
      }

      if (lastShift.length > 0) {
        return [
          ...shifts.slice(0, -1),
          [
            ...lastShift,
            shiftLine,
          ],
        ];
      }

      return shifts;
    }, [] as string[][]);

  return rawShifts
    .map((shiftLine) => getDataFromShiftLine(shiftLine, { zone, locale }));
};

export const getDataFromAdjustmentTable = (
  text: string[],
  lineBefore: string,
  excludedLabels: string[],
  lineBeforeOffset: number = 0,
): Adjustment[] => {
  const adjustments = chunk(text.slice(
    text.indexOf(lineBefore) + 1 + lineBeforeOffset,
  ), 2);

  return adjustments
    .map(([label, amount = '']) => ({
      label,
      amount: Math.round(Number.parseFloat(amount.slice(1)) * 100),
    }))
    .filter(({ label, amount }) => (
      !excludedLabels.includes(label)
      && !Number.isNaN(amount)
    ));
};

export const getPeriodFromHeader = (
  text: string[],
  lineLabel: string,
  labelSeparator: string,
  dateSeparator: string,
  dateTimeOpts: DateTimeOptions,
) => {
  const [start, end] = (text
    .find((line) => line.includes(lineLabel)) || '')
    .split(labelSeparator)[1]
    .split(dateSeparator)
    .map((date) => DateTime.fromFormat(date.trim(), INVOICE_DATE_FORMAT, dateTimeOpts));
  return {
    start,
    end: end.endOf('day'),
  };
};

export const getNameFromHeader = (text: string[], flag: string, offset: number = 0): string => {
  const flagIndex = text.findIndex((line) => line.includes(flag));
  const nameIndexOffset = text
    .slice(flagIndex + offset)
    .findIndex((line) => line.trim() !== '');
  return text[flagIndex + offset + nameIndexOffset].split(':')[1]
    .trim();
};
