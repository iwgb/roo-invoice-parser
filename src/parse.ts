import { Moment } from 'moment';
import hash from 'object-hash';
import { getPdfText, PdfData } from './utils/pdf';
import markets, { Markets } from './market/markets';

export interface Shift {
  start: Moment,
  end: Moment,
  orders: number,
  pay: number,
}

export interface Adjustment {
  label: string,
  amount: number,
}

export interface Invoice {
  start: Moment | null,
  end: Moment | null,
  shifts: Shift[],
  adjustments: Adjustment[],
  error: string,
  hash: string,
}

export interface InvoiceComponentGetterProps {
  text: string[],
  timezone: string,
}

const hashInvoice = (
  name: string,
  shifts: Shift[],
): string => hash({
  name,
  shifts: shifts.map((shift) => ({
    ...shift,
    start: shift.start.toISOString(),
    end: shift.end.toISOString(),
  })),
});

export const parseInvoice = async (
  data: PdfData,
  timezone: string,
  locale: keyof Markets,
): Promise<Invoice> => {
  const text = await getPdfText(data);

  let shifts: Shift[] = [];
  let adjustments: Adjustment[] = [];
  let period: {
    start: Moment | null,
    end: Moment | null,
  } = {
    start: null,
    end: null,
  };
  let name = '';
  let error = '';

  const parser = markets[locale];

  try {
    const args = { text, timezone };
    const [
      parsedName, parsedPeriod, parsedShifts, parsedAdjustments,
    ] = await Promise.all([
      parser.getName(args),
      parser.getPeriod(args),
      parser.getShifts(args),
      parser.getAdjustments(args),
    ]);
    name = parsedName;
    period = parsedPeriod;
    shifts = parsedShifts;
    adjustments = parsedAdjustments;
  } catch (e) {
    error = `${e.name}: ${e.message}`;
  }

  return {
    ...period,
    shifts,
    adjustments,
    error,
    hash: hashInvoice(name, shifts),
  };
};
