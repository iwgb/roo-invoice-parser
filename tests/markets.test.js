import fs from 'fs';
import asyncFs from 'fs/promises';
import path from 'path';
import { parseInvoiceText } from '../dist/parse';

const markets = fs.readdirSync(path.resolve(__dirname, './market/'));

markets.forEach((market) => {
  const marketName = market.split('.')[0];
  it(`should correctly parse ${marketName}`, async () => {
    const marketData = JSON.parse(await asyncFs.readFile(path.resolve(__dirname, './market/', market), 'utf-8'));
    expect(await parseInvoiceText(marketData))
      .toMatchSnapshot(marketName);
  });
});
