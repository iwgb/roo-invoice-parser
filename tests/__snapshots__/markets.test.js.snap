// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly parse en-GB: en-GB 1`] = `
Object {
  "adjustments": Array [
    Object {
      "amount": 0.5,
      "label": "Adjustments",
    },
  ],
  "currency": "GBP",
  "end": "2020-01-12T23:59:59.999+00:00",
  "error": "",
  "hash": "cb5f4b0283e52e204b0ef9e03328f8bb756c6a7d79f8f9dde872b7b32f38287e",
  "locale": "en-GB",
  "name": "Test Rider",
  "shifts": Array [
    Object {
      "end": "2020-01-06T20:43:00.000+00:00",
      "hours": 0.72,
      "orders": 1,
      "pay": 10,
      "start": "2020-01-06T20:00:00.000+00:00",
    },
    Object {
      "end": "2020-01-08T00:01:00.000+00:00",
      "hours": 4.02,
      "orders": 1,
      "pay": 10,
      "start": "2020-01-07T20:00:00.000+00:00",
    },
  ],
  "start": "2020-01-06T00:00:00.000+00:00",
}
`;
